File: src/chapters/differential-equations/content/code/heun.rs
// Heun's method (RK2).
fn heun<F: Fn(f64, f64) -> f64>(f: F, t0: f64, y0: f64, h: f64, n: usize) -> f64 {
    let (mut t, mut y) = (t0, y0);
    for _ in 0..n {
        let k1 = f(t, y); let k2 = f(t + h, y + h * k1);
        y += h * (k1 + k2) / 2.0; t += h;
    }
    y
}
fn main() {
    let f = |t: f64, y: f64| 2.0 * y - 10.0 * t * t + 2.0 * t;
    println!("y(1) = {}", heun(f, 0.0, 1.0, 0.1, 10));
}

File: src/pages/practice/guides/matlab/heun.m
function y = heun(f, t0, y0, h, n)
% HEUN  Heun's method (RK2, explicit trapezoidal rule).
    t = t0; y = y0;
    for i = 1:n
        k1 = f(t, y);
        k2 = f(t + h, y + h*k1);
        y = y + h*(k1 + k2)/2;
        t = t + h;
    end
end

% --- Demo ---
f = @(t, y) 2*y - 10*t^2 + 2*t;
fprintf('y(1) = %.6f\n', heun(f, 0, 1, 0.1, 10));
